File: backend/app/models/test_record.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, JSON
from sqlmodel import Field, Relationship, SQLModel

from .base import get_cst_now
from .user import User


# TestRecord - 测评记录
class TestRecordBase(SQLModel):
    test_name: str = Field(max_length=255)
    user_topic: str | None = Field(default=None, max_length=500)
    total_score: int | None = None
    total_max: int | None = None
    result_description: str | None = None
    questions: list = Field(default=[], sa_type=JSON)
    answers: list = Field(default=[], sa_type=JSON)
    scoring_ranges: list = Field(default=[], sa_type=JSON)
    conversation_id: str | None = None


class TestRecordCreate(TestRecordBase):
    pass


class TestRecordUpdate(SQLModel):
    test_name: str | None = Field(default=None, max_length=255)
    user_topic: str | None = Field(default=None, max_length=500)
    total_score: int | None = None
    total_max: int | None = None
    result_description: str | None = None
    scoring_ranges: list | None = None


class TestRecord(TestRecordBase, table=True):
    __tablename__ = "test_record"
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    created_at: datetime | None = Field(
        default_factory=get_cst_now,
        sa_type=DateTime(timezone=False),  # type: ignore
    )
    owner_id: uuid.UUID = Field(
        foreign_key="user.id", nullable=False, ondelete="CASCADE"
    )
    owner: User | None = Relationship(back_populates="test_records")


class TestRecordPublic(TestRecordBase):
    id: uuid.UUID
    created_at: datetime | None = None


class TestRecordsPublic(SQLModel):
    data: list[TestRecordPublic]
    count: int

File: backend-sb/src/main/java/com/emomind/entity/TestRecord.java
package com.emomind.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "test_record")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class TestRecord {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "test_name", nullable = false, length = 255)
    private String testName;

    @Column(name = "user_topic", length = 500)
    private String userTopic;

    @Column(name = "total_score")
    private Integer totalScore;

    @Column(name = "total_max")
    private Integer totalMax;

    @Column(name = "result_description", columnDefinition = "TEXT")
    private String resultDescription;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "questions", nullable = false, columnDefinition = "jsonb")
    private List<Map<String, Object>> questions;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "answers", nullable = false, columnDefinition = "jsonb")
    private List<Map<String, Object>> answers;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "scoring_ranges", columnDefinition = "jsonb")
    private List<Map<String, Object>> scoringRanges;

    @Column(name = "conversation_id")
    private String conversationId;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}
